packages/gestalt/src/NumberField.js

```diff
@@ -0,0 +1,152 @@
+   */
+  errorMessage?: Node,
+  /**
+   * More information for the user about how to complete the form field.
+   */
+  helperText?: string,
+  /**
+   * A unique identifier for the input.
+   */
+  id: string,
+  /**
+   * The label for the input. Be sure to localize the text.
+   */
+  label?: string,
+  /**
+   * The upper bound of valid input, inclusive.
+   */
+  max?: number,
+  /**
+   * The lower bound of valid input, inclusive.
+   */
+  min?: number,
+  /**
+   * A unique name for the input.
+   */
+  name?: string,
+  /**
+   * Callback triggered when the user blurs the input.
+   */
+  onBlur?: ({|
+    event: SyntheticFocusEvent<HTMLInputElement>,
+    value: number,
+  |}) => void,
+  /**
+   * Callback triggered when the value of the input changes, whether by keyboard entry or the input's arrows.
+   */
+  onChange: ({|
+    event: SyntheticInputEvent<HTMLInputElement>,
+    value: number,
+  |}) => void,
+  /**
+   * Callback triggered when the user focuses the input.
+   */
+  onFocus?: ({|
+    event: SyntheticFocusEvent<HTMLInputElement>,
+    value: number,
+  |}) => void,
+  /**
+   * Callback triggered when the user presses any key while the input is focused.
+   */
+  onKeyDown?: ({|
+    event: SyntheticKeyboardEvent<HTMLInputElement>,
+    value: number,
+  |}) => void,
+  /**
+   * Placeholder text shown the the user has not yet input a value.
+   */
+  placeholder?: string,
+  /**
+   * md: 40px, lg: 48px
+   */
+  size?: 'md' | 'lg',
+  /**
+   * Indicates the amount the value will increase or decrease when using the input's arrows.
+   */
+  step?: number,
+  /**
+   * The current value of the input.
+   */
+  value?: number | typeof undefined,
+|};
+
+/**
+ * [NumberField](https://gestalt.pinterest.systems/NumberField) allows for numerical input.
+ */
+const NumberFieldWithForwardRef: React$AbstractComponent<Props, HTMLInputElement> = forwardRef<
+  Props,
+  HTMLInputElement,
+>(function NumberField(
+  {
+    autoComplete,
+    disabled = false,
+    errorMessage,
+    helperText,
+    id,
+    label,
+    max,
+    min,
+    name,
+    onBlur,
+    onChange,
+    onFocus,
+    onKeyDown,
+    placeholder,
+    size = 'md',
+    step,
+    value,
+  }: Props,
+  ref,
+): Node {
+  return (
+    <InternalTextField
+      autoComplete={autoComplete}
+      disabled={disabled}
+      errorMessage={errorMessage}
+      helperText={helperText}
+      id={id}
+      label={label}
+      max={max}
+      min={min}
+      name={name}
+      onBlur={parseHandlerValue(onBlur)}
+      onChange={parseHandlerValue(onChange)}
+      onFocus={parseHandlerValue(onFocus)}
+      onKeyDown={parseHandlerValue(onKeyDown)}
+      placeholder={placeholder}
+      size={size}
+      step={step}
+      ref={ref}
+      type="number"
+      // See comment above — we need to stringify what we give InternalTextField
+      value={value === undefined ? value : String(value)}
```

Comment:
nit

value === undefined ? value : String(value)
vs
value === undefined ? undefined : String(value)

?

Reply:
It doesn't really matter, but I like the philosophy of "we're always returning the value, just sometimes we stringify it" as opposed to "sometimes we return the value, sometimes we return something else" — even though in practice it doesn't change anything here. It just seems more "correct" to me philosophically.